File: ♦♦2370-LongestIdealSubsequence.py
class Solution:
    def longestIdealString(self, s: str, k: int) -> int:
        n = len(s)
        dp = [0] * n
        last_occurrence = [-1] * 26

        for i in range(n):
            dp[i] = 1
            for j in range(max(0, ord(s[i]) - ord('a') - k), min(26, ord(s[i]) - ord('a') + k + 1)):
                if last_occurrence[j] != -1:
                    dp[i] = max(dp[i], dp[last_occurrence[j]] + 1)
            last_occurrence[ord(s[i]) - ord('a')] = i

        return max(dp)

File: ♦♦2370-LongestIdealSubsequence.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */

var longestIdealString = function(s, k) {
    let n = s.length;
    let dp = new Array(n).fill(0);
    let last_occurrence = new Array(26).fill(-1);

    for (let i = 0; i < n; i++) {
        dp[i] = 1;
        for (let j = Math.max(0, s.charCodeAt(i) - 'a'.charCodeAt(0) - k); j <= Math.min(25, s.charCodeAt(i) - 'a'.charCodeAt(0) + k); j++) {
            if (last_occurrence[j] != -1) {
                dp[i] = Math.max(dp[i], dp[last_occurrence[j]] + 1);
            }
        }
        last_occurrence[s.charCodeAt(i) - 'a'.charCodeAt(0)] = i;
    }

    return Math.max(...dp);
};
